src/interchange/avro.rs

```diff
@@ -412,86 +494,42 @@ impl Decoder {
             None => (&self.reader_schema, None),
         };
 
-        fn value_to_datum(v: &Value) -> Result<Datum<'_>> {
-            match v {
-                Value::Null => Ok(Datum::Null),
-                Value::Boolean(true) => Ok(Datum::True),
-                Value::Boolean(false) => Ok(Datum::False),
-                Value::Int(i) => Ok(Datum::Int32(*i)),
-                Value::Long(i) => Ok(Datum::Int64(*i)),
-                Value::Float(f) => Ok(Datum::Float32((*f).into())),
-                Value::Double(f) => Ok(Datum::Float64((*f).into())),
-                Value::Date(d) => Ok(Datum::Date(*d)),
-                Value::Timestamp(d) => Ok(Datum::Timestamp(*d)),
-                Value::Decimal { unscaled, .. } => Ok(Datum::Decimal(
-                    Significand::from_twos_complement_be(&unscaled)?,
-                )),
-                Value::Bytes(b) => Ok(Datum::Bytes(b)),
-                Value::String(s) => Ok(Datum::String(s)),
-                Value::Union(v) => value_to_datum(v),
-                other @ Value::Fixed(..)
-                | other @ Value::Enum(..)
-                | other @ Value::Array(_)
-                | other @ Value::Map(_)
-                | other @ Value::Record(_) => bail!("unsupported avro value: {:?}", other),
-            }
-        };
-
-        fn extract_row<'a, I>(v: Value, extra: I) -> Result<Option<Row>>
-        where
-            I: IntoIterator<Item = Datum<'a>>,
-        {
-            let v = match v {
-                Value::Union(v) => *v,
-                _ => bail!("unsupported avro value: {:?}", v),
-            };
-            match v {
-                Value::Record(fields) => {
-                    let mut row = RowPacker::new();
-                    for (_, col) in fields.iter() {
-                        row.push(value_to_datum(col)?);
-                    }
-                    for d in extra {
-                        row.push(d);
-                    }
-                    Ok(Some(row.finish()))
-                }
-                Value::Null => Ok(None),
-                _ => bail!("unsupported avro value: {:?}", v),
+        let result = if self.is_debezium {
+            if let (Some(schema), None) = (&self.fast_row_schema, reader_schema) {
+                // The record is laid out such that we can extract the `before` and
+                // `after` fields without decoding the entire record.
+                let before = extract_row(
+                    block_on(avro_rs::from_avro_datum(&schema, &mut bytes, None))?,
```

Comment:
as discussed this is fine as is, but it does seem a tiny bit unfortunate that we went through all the work to turn from_avro_datum async and then just call `block_on` on it.

Reply:
Yep -- but this is only called in the Kafka path. We are genuinely using the asynchronicity in the OCF path.

I'll open an issue for really beating on Avro performance since, as we discussed offline, even before this change there is a lot of possible improvement to be made.